cli/tests/program.rs

```diff
@@ -1614,3 +1825,176 @@ fn test_cli_program_dump() {
+    config: &mut CliConfig<'a>,
+    online_signer: &'a Keypair,
+    offline_signer: &'a Keypair,
+    buffer_signer: &'a Keypair,
+    buffer_signer_identity: &'a NullSigner,
+) {
+    // Write a buffer
+    config.signers = vec![online_signer, buffer_signer];
+    config.command = CliCommand::Program(ProgramCliCommand::WriteBuffer {
+        program_location: program_path.to_str().unwrap().to_string(),
+        fee_payer_signer_index: 0,
+        buffer_signer_index: Some(1),
+        buffer_pubkey: Some(buffer_signer.pubkey()),
+        buffer_authority_signer_index: 0,
+        max_len: None,
+        skip_fee_check: false,
+    });
+    process_command(config).unwrap();
+    let buffer_account = rpc_client.get_account(&buffer_signer.pubkey()).unwrap();
+    if let UpgradeableLoaderState::Buffer { authority_address } = buffer_account.state().unwrap() {
+        assert_eq!(authority_address, Some(online_signer.pubkey()));
+    } else {
+        panic!("not a buffer account");
+    }
+
+    // Set buffer authority to offline signer
+    config.signers = vec![online_signer, buffer_signer];
+    config.command = CliCommand::Program(ProgramCliCommand::SetBufferAuthority {
+        buffer_pubkey: buffer_signer.pubkey(),
+        buffer_authority_index: Some(0),
+        new_buffer_authority: offline_signer.pubkey(),
+    });
+    config.output_format = OutputFormat::JsonCompact;
+    let response = process_command(config);
+    let json: Value = serde_json::from_str(&response.unwrap()).unwrap();
+    let offline_signer_authority_str = json
+        .as_object()
+        .unwrap()
+        .get("authority")
+        .unwrap()
+        .as_str()
+        .unwrap();
+    assert_eq!(
+        Pubkey::from_str(offline_signer_authority_str).unwrap(),
+        offline_signer.pubkey()
+    );
+    let buffer_account = rpc_client.get_account(&buffer_signer.pubkey()).unwrap();
+    if let UpgradeableLoaderState::Buffer { authority_address } = buffer_account.state().unwrap() {
+        assert_eq!(authority_address, Some(offline_signer.pubkey()));
+    } else {
+        panic!("not a buffer account");
+    }
+
+    // Attempt to deploy from buffer using previous authority (should fail)
+    config.signers = vec![online_signer, buffer_signer_identity];
+    config.command = CliCommand::Program(ProgramCliCommand::Deploy {
+        program_location: None,
+        fee_payer_signer_index: 0,
+        program_signer_index: None,
+        buffer_signer_index: Some(1),
+        allow_excessive_balance: false,
+        upgrade_authority_signer_index: 0,
+        is_final: false,
+        max_len: None,
+        skip_fee_check: false,
+    });
+    config.output_format = OutputFormat::JsonCompact;
+    let error = process_command(config).unwrap_err();
+    assert_eq!(error.to_string(), "Deploying program failed: RPC response error -32002: Transaction simulation failed: Error processing Instruction 1: Incorrect authority provided [5 log messages]");
```

Comment:
This is doing more than just creating the buffer by trying the deploy. This behavior is checked in `test_cli_program_mismatch_buffer_authority()`, so we can remove doing it here

Reply:
True, I do prefer more checks when possible (and I'm typically lazy to cross check against every existing test whether or not this particular behavior has been tested already or not),

that aside, here I'm actually **checking for a somewhat different thing** compared to what `test_cli_program_mismatch_buffer_authority` is doing:
- `test_cli_program_mismatch_buffer_authority` is trying to supply "fake" authority that has never been this buffer's authority in the past
- while in `create_buffer_with_offline_authority` I'm essentially testing that "old" authority has no control over our buffer anymore

Reply:
It's typically more of a maintenance burden to have copy-pasta checks in unrelated tests.

If you feel strongly about it, I recommend adding the old authority failure check in `test_cli_program_set_buffer_authority` in a separate PR. What do you think?

Reply:
Seems like an important thing to check for (it works as intended now, but having a "net of tests" to catch regressions is probably also important for mature codebase),

moved this check into `test_cli_program_set_buffer_authority` (https://github.com/solana-labs/solana/pull/33860/commits/a671a2c3643ff3b575f805d309cb3094e3023a00) in a [separate PR](https://github.com/solana-labs/solana/pull/34316).